### DISCUSSION TITLE
Configuring Scram-sha-512 and tls authentication for kafka cluster

### DISCUSSION THREAD

**Discussion by Lokeshmaxi** [strimzi/strimzi-kafka-operator#9349] [Q&A]
### Suggestion / Problem

Configuring Scram-sha-512 and tls authentication for kafka cluster.
I have deployed strimzi kafka cluster in kubernetes cluster, we need to enable authentication for connecting bootstrap server from the client, so i have tried and configured SCRAM-SHA-512 authentication with authorization type "simple"   and created one user with all privilege(for testing), now my question is, how to pass client properties which consist of SCRAM-SHA-512,Username and pw in the kafka cluster inside the pod, there is Vi editor inside the kafka container, and  is creating client properties need to everytime when pod restarts?,is there way to pass it in within config properties of kafka, and one more help  is there any detailed document to implement to authentication and authorization for all type for strimzi kafka, we are planning implement kafka for our production environment soon, have attached my configuration screenshot as reference, please help

Regards
Lokeshkumar

![Kafkauser](https://github.com/strimzi/strimzi-kafka-operator/assets/91049865/954131ae-598e-4a4e-9a59-26e7c4979036)
![Kafka-cluster](https://github.com/strimzi/strimzi-kafka-operator/assets/91049865/b0fafd9a-b010-4e54-bf96-b4950f47a7c7)

**Comment by scholzj** [comment]
I'm not entirely sure what the question is. You should never run your Kafka client from inside the broker pods. That can cause problems for example because the container might run out of memory, might degrade performance etc. When you create the `KafkaUser` resource, a Kubernetes Secret is created with the credentials. You can use that in your Kafka client in any way you want - and in any way the client or programming language you use needs.

**Reply by Lokeshmaxi** [reply]
post that kafkauser creation, i have  executed kafka pod and tried to create topic using kafka cli, but it is throwing Authorization denied, my question is that we can also able to create topic using  the kafka cli within the pod right ?

**Reply by scholzj** [reply]
You should never execute into the KAfka broker pods to run some command. Start your own pod and run it from there. In any case, if you get authorization error:
* Make sure you are authenticated (in your config, the port 9092 has no authentication, so it will likely always get "Not authorized" error)
* Check the broker log to see what exactly is the ACL right you are missing.

**Reply by Lokeshmaxi** [reply]
I have configured authorization from 9094 to 9092, and tried to create topic, but getting this error from kafka pod=listenerType=ZK_BROKER Failed authentication (Unexpected Kafka request of type METADATA during SASL handshake.) (org.apache.kafka.common.network.Selector) [data-plane-kafka-network-thread-0-ListenerName(PLAIN-9092)-SASL_PLAINTEXT-5]

**Comment by scholzj** [comment]
Well, that suggests you have something wrong with the authenication setup
of your client.

**Reply by Lokeshmaxi** [reply]
Hi Schokzj, can you please share the some other document to implement Scram-sha-512 for authentication for kafka, i have tried many times, still it is not working, please guide me

**Comment by Lokeshmaxi** [comment]
`
![Configured scram-sha-512](https://github.com/strimzi/strimzi-kafka-operator/assets/91049865/d6bb7d78-2e10-4a8f-b76a-37efeb7db41a)


Attached` Scram-sha-512 configuration to kafka, please let me know is this correct or not

**Reply by scholzj** [reply]
This is not how you configure Strimzi. All the options around security you use there are forbidden and will not be used for anything. If you want to configure authentication, you have to do it in the `.spec.kafka.listeners.authentication` section of the `Kafka` custom resource. There are examples in the docs and on our GitHub. The same applies to authorization.
